### ISSUE TITLE
Change default branch name?

### ISSUE THREAD

**Issue by henry-ngo** [bcgov/bcgovr#56]
This quoted line below assumes that the user has named their default branch on GitHub "master", which isn't always the case. Especially with the @repomountie bot guidelines on master and main branch naming (i.e. both are suggested choices). Perhaps choose one as the "default" but allow user to specify a different branch name with a keyword argument?

https://github.com/bcgov/bcgovr/blob/affe8ce20286a6e53cb11be0a6e0aec1027eb7cd/R/use.R#L208

Note: There is a similar phrasing in the template CONTRIBUTING.md file generated too. Is there a branch-name-agnostic way to rephrase: 

> Pull requests will be evaluated by the repository guardians on a schedule and if deemed beneficial will be committed to the master.

Thanks for your consideration!

**Comment by ateucher** [comment]
This is a great catch, thanks @henry-ngo. We are planning on changing the default branch for all the repos that we can, so this would bite eventually bite us (and other users). Also good call on the `CONTRIBUTING.md` template.

**Comment by boshek** [comment]
Replacing `master` with `main`?

**Comment by ateucher** [comment]
For the code, we should get it dynamically using the GitHub API:

    library(gh)
    gh("/repos/bcgov/bcgovr")$default_branch
    #> [1] "master"

And in the `CONTRIBUTING.md` file, just be generic and refer to it as the "default branch"

**Comment by henry-ngo** [comment]
Generic reference to default branch sounds great to me :)

I think getting dynamic names is a great idea but I don't think the suggested lines will work. This is fairly new to me so sorry if I'm just making a mistake, but I believe the git push command being used here is being called to initialize the GitHub repo. In my mind,  the user already has git initialized for their project at the local level and now they are creating a GitHub repo for their project in the BCGOV organization and pushing a local branch to the remote repo (currently the local branch named `master`). 

So, getting the **GitHub** default branch name won't work, we need to be using the default local branch name. The user would have had to run `use_bcgov_git()` first, and that source code doesn't seem to specify any branch names so that would default to whatever each user has set as their own default branch name for new repositories.  But maybe I'm wrong about this?? If so, then just ignore the rest of this comment :)

I think changing that line to just be `git push -u origin` might work since leaving out the branch name could in theory just cause the user to push **all** of their local branches, which would be only their default branch if that's the only branch created so far. But this may not be desirable if they have already made several branches locally for development and only want to put the main branch on GitHub. In addition, the behaviour of `git push` without a branch name specified is variable and depends on user configuration of git, so this can be a dangerous and unpredictable choice. 

Not sure what's the best choice since I'm not familiar enough with this package's general philosophy on these choices! The shortcut above where the branch name is left out is probably going to work most of the time but it can cause a lot of problems if it fails by pushing all branches unexpectedly. Perhaps there's a way to figure out what the current branch is and use that (main downside: if the user is on a developmental branch when they run this then it wouldn't be ideal). 

Maybe the most foolproof solution but more code needed: Add optional named argument for branch name. Use that value if provided. Otherwise, check if there's a branch named `main` and use that and if not, then check for `master` and use that, and if not, then fail and request user to specify a branch?

**Comment by boshek** [comment]
If we are talking about changing the default branch name of _new_ projects we could add something like this:

`system('git branch -m master main')`
 
to here:

https://github.com/bcgov/bcgovr/blob/fba0050b95490d5ec9f4c9e7c7c129b0ce3c6209/R/use.R#L231

If we wanted to get more spiffy we directly use `git2r` and rename it that way. I'm not sure bcgovr is the place for a function to rename existing branches.

**Comment by ateucher** [comment]
This is relevant: https://github.com/r-lib/usethis/issues/1156

**Comment by ateucher** [comment]
@henry-ngo did you actually run into this, or just notice it in the source code? Remembering back, this whole [`tryCatch` block](https://github.com/bcgov/bcgovr/blob/affe8ce20286a6e53cb11be0a6e0aec1027eb7cd/R/use.R#L203-L215) was just there to [work around](https://github.com/bcgov/bcgovr/commit/60f8bc43e6b85853153c89a76c2a98a2e8edb32c) a bug in usethis/git2r, which I'm pretty was fixed long ago. I'm wondering if we can just get rid of it and rely on usethis, which I think should/does/will simply respect a user's git configuration (eg., the new [`init.defaultBranch` config setting](https://git-scm.com/docs/git-config#Documentation/git-config.txt-initdefaultBranch))?

Regarding suggestions of renaming branches, I think we want to avoid that.

**Comment by henry-ngo** [comment]
Sorry to not have been clear! I did run into this when I tried to use this function to initialize the github repo for https://github.com/bcgov/epi.branch.sim In the end, I just did it manually :)

**Comment by ateucher** [comment]
Ok, thanks. Perhaps we still do need that catch then. Just to make sure I understand, you hadn't created an empty repo on GitHub beforehand, right?

**Comment by henry-ngo** [comment]
Good point. Sadly, I don't remember right now if I tried to create an empty repo and then push via this `bcgovr` function first (apparently incorrect) or if I created an empty repo after the function failed. I could try to replicate this with a test repo if you'd like (but I might need find a bit of time to do that later this week).

**Comment by boshek** [comment]
Ok so the solution with respect to the `use_bcgov_git()` here to is follow the lead of `usethis`? Seems like their approach is sound.

**Comment by ateucher** [comment]
Yes I think so... and I think maybe we should remove that tryCatch anyway, the error is between user and `usethis`, maybe we don't want to insert a fragile fix in the middle?
